jdaviz/configs/default/plugins/export_plot/export_plot.py

```diff
@@ -74,3 +85,88 @@ def vue_save_figure(self, filetype):
         the bqplot.Figure save methods.
         """
         self.save_figure(filetype=filetype)
+
+    def save_movie(self, i_start, i_end, filename=None, filetype=None, rm_temp_files=True):
+        """``i`` start/end control the frames being written out.
+        ``i_end`` is inclusive. This method creates a bunch of
+        PNG files (one per frame) and then deletes them after stitching
+        the video.
+        """
+        if not HAS_OPENCV:
+            raise ImportError("Please install opencv-python to save cube as movie.")
+
+        if self.config != "cubeviz":
+            raise NotImplementedError(f"save_movie is not available for config={self.config}")
+
+        if filetype is None:
+            if filename is not None and '.' in filename:
+                filetype = filename.split('.')[-1]
+            else:
+                # default to MPEG-4
+                filetype = "mp4"
+
+        if filetype != "mp4":
+            raise NotImplementedError(f"filetype={filetype} not supported")
+
+        viewer = self.viewer.selected_obj
+        if viewer.shape is None:
+            raise ValueError("Selected viewer has no display shape.")
+
+        if filename is None:
+            filename = f"mymovie.{filetype}"
+
+        slice_plg = self.app._jdaviz_helper.plugins["Slice"]._obj
+        orig_slice = slice_plg.slice
+        ts = float(slice_plg.play_interval) * 1e-3  # ms to s
+        temp_png_files = []
+
+        # TODO: Expose to users?
+        frame_size = viewer.shape[::-1]  # nx, ny
+        fps = 5  # This is arbitrary
+        i_step = 1  # Need n_frames check if we allow tweaking
+
+        # No wrapping. Forward only.
+        if i_start < 0:
+            i_start = 0
+        if i_end > slice_plg.max_value:
+            i_end = slice_plg.max_value
+        if i_end <= i_start:
+            raise ValueError(f"No frames to write: i_start={i_start}, i_end={i_end}")
+
+        video = cv2.VideoWriter(filename, cv2.VideoWriter_fourcc(*'mp4v'), fps, frame_size, True)
+        i = i_start
+
+        try:
+            self.app.loading = True  # Grays out the app
+
+            while i <= i_end:
+                slice_plg._on_slider_updated({'new': i})
+                cur_pngfile = f"._cubeviz_movie_frame_{i}.png"
+
+                # FIXME: bqplot stuck after first frame!
+                # If we can fix this, maybe we can have callback to video.write and don't need PNG files.
+                self.save_figure(filename=cur_pngfile, filetype="png")

```

Comment:
@mariobuikhuizen or @maartenbreddels , I don't understand why `figure.get_png_data()` gets stuck after the first frame is written out. It never seems to get into `finally`. I tried to brute force setting `figure._upload_png_callback = None` but that did not help at all. What is going on here? 🙏 

```python
from jdaviz import Cubeviz

cubeviz = Cubeviz()
cubeviz.load_data(...)
cubeviz.show()

plg = cubeviz.plugins["Export Plot"]
plg._obj.save_movie(1000, 1010, filename="zzzz.mp4", rm_temp_files=False)
```

```
----> 1 plg._obj.save_movie(1000, 1010, filename="zzzz.mp4", rm_temp_files=False)

File jdaviz/configs/default/plugins/export_plot/export_plot.py:145, in ExportViewer.save_movie(self, i_start, i_end, filename, filetype, rm_temp_files)
    143 slice_plg._on_slider_updated({'new': i})
    144 cur_pngfile = f"._cubeviz_movie_frame_{i}.png"
--> 145 self.save_figure(filename=cur_pngfile, filetype="png")
    146 temp_png_files.append(cur_pngfile)
    147 i += i_step

File jdaviz/configs/default/plugins/export_plot/export_plot.py:76, in ExportViewer.save_figure(self, filename, filetype)
     74             with open(os.path.expanduser(filename), 'bw') as f:
     75                 f.write(data)
---> 76         viewer.figure.get_png_data(on_img_received)
     77 elif filetype == "svg":
     78     viewer.figure.save_svg(filename)

File bqplot/figure.py:207, in Figure.get_png_data(self, callback, scale)
    195 '''
    196 Gets the Figure as a PNG memory view
    197 
   (...)
    204     Scale up the png resolution when scale > 1, when not given base this on the screen pixel ratio.
    205 '''
    206 if self._upload_png_callback:
--> 207     raise Exception('get_png_data already in progress')
    208 self._upload_png_callback = callback
    209 self.send({'type': 'upload_png', 'scale': scale})

Exception: get_png_data already in progress
```

Reply:
Thanks!